### PR TITLE
FIP-19: Add RAI Bonding Curve

### PR DESCRIPTION
**Pull request by guifel**
This Proposal adds a RAI bonding curve with the following parameters:

scale - 6M
discount pre-scale - 0.3%
premium post-scale - 2%
allocation - 100% to Fuse pool 9

Access control changes:
Minter - RAI Bonding Curve

### CODE REVIEW

**Comment by guifel** [comment]
Left todo:
- [ ] Set the mint cap to 10m FEI
- [ ] Deploy contracts, run test and generate call data
- [ ] Add the the 1m FEI supply to Fuse pool 9

File: end-to-end/e2e.spec.ts [hunk 1]
```diff
@@ -133,25 +133,59 @@ describe('e2e', function () {
...
         await bondingCurve.allocate()
       
         const curveBalanceAfter = await bondingCurve.balance();
         expect(curveBalanceAfter).to.be.bignumber.equal(toBN(0))
-        
-        const pcvDepositAfter = await uniswapPCVDeposit.balance()
-        await expectApprox(pcvDepositAfter, allocatedDpi.mul(toBN(9)).div(toBN(10)), '100')
 
-        await expectApprox(await fusePCVDeposit.balance(), allocatedDpi.div(toBN(10)), '100')
+        await expectApprox(await fusePCVDeposit.balance(), allocatedRai.div(toBN(10)), '100')
```
Comment (Joeysantoro) [inline end-to-end/e2e.spec.ts:188]: Should probably remove the div(10) here as 100% of the rewards go to Fuse

**Review by Joeysantoro — COMMENTED** [review]
You'll want to add fip_18 to proposals/config.json so that the end-to-end tests can run. Something like 
```
"fip_18" : {
"deploy" : true,
"execute" : false
}
```
should work. Then you can run `npm run test:e2e` to see if the integration tests pass

**Comment by guifel** [comment]
I could run the e2e tests. RAI bonding curve tests are now passing.

I added a 10m mint cap, please check: 4849ccdd10b101ed6772f42347e31d82e147c36c

So what left to be done is deploying the contracts, adding the addresses and fill-in the calldata in proposal/description/fip_19.json

Note that the reweight test fails but it also fails on the current master branch:
```
  1) e2e
       Reweights
         should perform reweight above peg correctly:

      AssertionError: expected '2894' to equal '2895'

```

File: proposals/description/fip_19.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    "proposal_commands": [
+        {
+            "address": "0x8d5ED43dCa8C2F7dFB20CF7b53CC7E593635d7b9",
+            "values": "0",
+            "method": "grantMinter(address)",
+            "arguments": [
+                "TODO Add raiBondingCurveAddress"
+            ],
+            "description": "Grant Minter to RAI Bonding Curve"
+        }
```
Comment (Joeysantoro) [inline proposals/description/fip_19.json:13]: Don't forget to add the mint cap raising step here

**Review by Joeysantoro — COMMENTED** [review]
Overall looks good

File: contract-addresses/mainnetAddresses.json [hunk 3]
```diff
@@ -9,12 +9,15 @@
...
     "daiBondingCurveAddress": { "artifact": "BondingCurve", "address":  "0xC0afe0E649e32528666F993ce63822c3840e941a"},
     "compositeOracleAddress": { "artifact": "CompositeOracle", "address":  "0x8721f9EAba0B9081069970bCBce38763D3D4f28E"},
     "chainlinkEthUsdOracleWrapperAddress": { "artifact": "ChainlinkOracleWrapper", "address":  "0xCd3c40AE1256922BA16C7872229385E20Bc8351e"},
     "chainlinkFeiEthOracleWrapperAddress": { "artifact": "ChainlinkOracleWrapper", "address":  "0x060Be7B51F78DFFd04749332fd306BA1228e7444"}, 
     "chainlinkDpiUsdOracleWrapperAddress": { "artifact": "ChainlinkOracleWrapper", "address":  "0xB594d2bd55Ede471e16b92AE6F7651648DA871c3"}, 
     "chainlinkDaiUsdOracleWrapperAddress": { "artifact": "ChainlinkOracleWrapper", "address":  "0x231aDa12E273eDf3fA54CbD90c5C1a73129D5bb9"},
+    "chainlinkRaiEthOracleWrapperAddress": { "artifact": "ChainlinkOracleWrapper", "address":  "0x3d49573ee6aFCBDe606F8a1c2AA1C498048E7190"},
+    "chainlinkRaiUsdCompositOracleAddress": { "artifact": "CompositeOracle", "address":  "0x392b1d29eDab680c5CA778D3A32b8284859BFBB0"},
```
Comment (Joeysantoro) [inline contract-addresses/mainnetAddresses.json:20]: verify on etherscan

File: contract-addresses/mainnetAddresses.json [hunk 4]
```diff
@@ -83,6 +90,8 @@
     "rariPool8EthIrmAddress": {"address": "0xbab47e4b692195bf064923178a90ef999a15f819" },
     "rariPool8DaiIrmAddress": {"address": "0xede47399e2aa8f076d40dc52896331cba8bd40f7" },
     "aWETHAddress" : {"address": "0x030bA81f1c18d280636F32af80b9AAd02Cf0854e"},
+    "aRaiAddress" : {"address": "0xc9BC48c72154ef3e5425641a3c747242112a46AF"},
+    "aaveRaiPCVDeposit": {"artifact" : "AavePCVDeposit", "address": "0xd2174d78637a40448112aa6B30F9B19e6CF9d1F9"},
```
Comment (Joeysantoro) [inline contract-addresses/mainnetAddresses.json:94]: verify on etherscan

**Review by Joeysantoro — APPROVED** [review]
🚢

**Pull request merged**